org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/CompletionEngineProviderDiscovery.java

```diff
@@ -0,0 +1,91 @@
+/*******************************************************************************
+ * Copyright (c) 2025 Red Hat, Inc. and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.jdt.internal.core;
+
+import java.util.Objects;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IConfigurationElement;
+import org.eclipse.core.runtime.IExtension;
+import org.eclipse.core.runtime.IExtensionPoint;
+import org.eclipse.core.runtime.ILog;
+import org.eclipse.core.runtime.Platform;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.internal.codeassist.CompletionEngine;
+import org.eclipse.jdt.internal.codeassist.ICompletionEngineProvider;
+
+class CompletionEngineProviderDiscovery {
+	private static final String SELECTED_SYSPROP = "ICompletionEngineProvider"; //$NON-NLS-1$
+	private static final String CODE_ASSIST_PROVIDER_EXTPOINT_ID = "completionEngineProvider" ; //$NON-NLS-1$
+	private static boolean ERROR_LOGGED = false;
+
+	private static String lastId;
+	private static IConfigurationElement lastExtension;
+
+	public static ICompletionEngineProvider getInstance() {
+		String id = System.getProperty(SELECTED_SYSPROP);
+		IConfigurationElement configElement = getConfigurationElement(id);
+		lastId = id;
+		lastExtension = configElement;
+		if (configElement != null) {
+			try {
+				// We do prefer creating a new instance on each call, as it can allow the extension
+				// to store some state more easily than by using a singleton.
```

Comment:
I would maybe test roughly how long this part takes in regular usage. I would have thought caching after the first invocation would be a way to avoid any surprises.

Reply:
On second though, I guess this isn't so bad because in the normal case, there's no extension defined for the standard use case, so none of that runs.

Reply:
Takes around a millisecond. I could do some caching with a HashMap to avoid recreating it. To be honest, I copied this code from the CompilationUnitResolver extension point, so we should be able to cache it safely

Reply:
I've implemented a caching mechanism